fix the required + nullable issue

ArcturusZhang opened the pull request:
Fixes https://github.com/Azure/autorest.csharp/issues/2340
Fixes https://github.com/Azure/autorest.csharp/issues/3363
Fixes https://github.com/Azure/autorest.csharp/issues/3440
Fixes https://github.com/Azure/autorest.csharp/issues/3600

# Description

This PR fixes multiple issues around nullable properties, especially when the property is required.

This PR also updates the `InputType` to make `IsNullable` property no longer optional. This optional property `IsNullable` could be problematic because of its optionality, sometimes we forget to give it a value and it causes hidden issues.

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

ArcturusZhang commented on test/CadlRanchProjects/type/property/nullable/src/Generated/Models/CollectionsByteProperty.cs:
```diff
@@ -18,14 +18,13 @@ public partial class CollectionsByteProperty
         /// <param name="nullableProperty"> Property. </param>
-        /// <exception cref="ArgumentNullException"> <paramref name="requiredProperty"/> or <paramref name="nullableProperty"/> is null. </exception>
+        /// <exception cref="ArgumentNullException"> <paramref name="requiredProperty"/> is null. </exception>
         internal CollectionsByteProperty(string requiredProperty, IEnumerable<BinaryData> nullableProperty)
```
This corresponds to the typespec:
```
model CollectionsByteProperty {
nullableProperty: bytes[] | null;
}
```
But in our design, an array should never be null. Should we change the behavior here back?

m-nash replied:
Good catch let me get some more information on this.

m-nash commented on test/CadlRanchProjects/type/property/nullable/src/Generated/Models/CollectionsByteProperty.cs:
```diff
@@ -18,14 +18,13 @@ public partial class CollectionsByteProperty
 
             RequiredProperty = requiredProperty;
-            NullableProperty = nullableProperty.ToList();
+            NullableProperty = nullableProperty?.ToList();
```
This needs to be initialized as a ChangeTrackingList if its `nullable` or `optional`.  If we use Optional<T> here the .ToList takes care of that for us.

ArcturusZhang replied:
I agree that we should initialize this as a ChangeTrackingList and never have a list to be null anywhere
But here we cannot use Optional.ToList because this is in the public constructor. The `Optional<T>` only appear in the deserialization method.

ArcturusZhang commented:
The assignment for nullable arrays is breaking test cases like this: https://github.com/Azure/autorest.csharp/blob/0df9bf7948af9d87cd8ff18071465cef5d665e3e/test/AutoRest.TestServer.Tests/ModelShapeTests.cs#L183

ArcturusZhang commented on test/TestProjects/ModelShapes/Generated/Models/InputModel.Serialization.cs:
```diff
@@ -81,7 +81,7 @@ void IUtf8JsonSerializable.Write(Utf8JsonWriter writer)
                 writer.WriteNull("RequiredNullableInt");
             }
-            if (RequiredNullableStringList != null)
+            if (RequiredNullableStringList != null && Optional.IsCollectionDefined(RequiredNullableStringList))
```
I have to add this additional check to check undefined in order to write `null`

pshao25 commented on test/CadlRanchProjects/type/property/nullable/src/Generated/Models/CollectionsByteProperty.Serialization.cs:
```diff
@@ -32,12 +32,17 @@ internal static CollectionsByteProperty DeserializeCollectionsByteProperty(JsonE
                         if (item.ValueKind == JsonValueKind.Null)
                         {
-                            array.Add(null);
+                            continue;
```
Just continue? Looks not correct to me. I thought either throw error or just add null to array.

ArcturusZhang replied:
we should never throw

ArcturusZhang replied:
I will revert this

pshao25 commented:
Let's remove all the Ignore in test\CadlRanchProjects.Tests\type-property-nullable.cs to see if it passes.

pshao25 commented:
Could you add tests in cadl-ranch for cases that the element of array is nullable and non-nullable?

ArcturusZhang commented:
Current behavior for required nullable arrays:

# API and serialization side

Schema:
```
model Model
{
array: T[] | null;
}
```
where `T` is either a reference type or value type, we have no difference.

Generated code:
```csharp
public class Model
{
      // public constructor
      public Model(IEnumerable<T> array)
      {
           Array = array?.ToList() as IList<T> ?? new ChangeTrackingList<T>(); // ensure this is never null
      }
}
```

Serialization:
- if `null` is passed into the ctor, `var model = new Model(null)`, after serialization, we have:
```
{
    "array": null
}
```
- if empty array is passed into the ctor, `var model = new Model(Array.Empty<T>())`, after serialization, we have:
```
{
    "array": []
}
```

# Deserialization side

Here we have differences between value types and reference types, but in general, in the DeserializationModel static method, in the last step that it is calling the internal ctor, we will create an instance of the ChangeTrackingList to ensure it is still not null: https://github.com/Azure/autorest.csharp/blob/fa50b20e0e86d40adec0d7dd7b87f3aeae4b7eb7/test/CadlRanchProjects/type/property/nullable/src/Generated/Models/CollectionsByteProperty.Serialization.cs#L56

In the deserializing property part, for int32 and other null-unaccepting element types
```
model Model
{
array: int32[] | null;
}
```
We get code like this:
```csharp
List<int> array = new List<int>();
foreach (var item in property.Value.EnumerateArray())
{
    array.Add(item.GetInt32());
}
```
if there is a `null` in the array in the response, we will get an exception.

if the element is a reference type, we have the following code:
```csharp
List<BinaryData> array = new List<BinaryData>();
foreach (var item in property.Value.EnumerateArray())
{
    if (item.ValueKind == JsonValueKind.Null)
    {
        array.Add(null);
    }
    else
    {
        array.Add(BinaryData.FromBytes(item.GetBytesFromBase64("D")));
    }
}
```
if there is a `null` in the array in the response, we will add it into the array.

if the element is a complex model, we have the following code:
```csharp
List<CollectionItem> array = new List<CollectionItem>();
foreach (var item in property.Value.EnumerateArray())
{
    array.Add(CollectionItem.DeserializeCollectionItem(item));
}
```
if there is a `null` in the array in the response, we will add it into the array.

ArcturusZhang commented on test/TestProjects/ModelShapes/Generated/Models/InputModel.Serialization.cs:
```diff
@@ -81,34 +81,20 @@ void IUtf8JsonSerializable.Write(Utf8JsonWriter writer)
-                writer.WriteStartArray();
-                foreach (var item in RequiredNullableIntList)
-                {
-                    writer.WriteNumberValue(item);
-                }
-                writer.WriteEndArray();
+                writer.WriteStringValue(item);
             }
-            else
+            writer.WriteEndArray();
```
Need to investigate this - this actually changes the serialization result

ArcturusZhang commented:
I am finding more and more issues when I am trying to "ensure the collection property can never be null".
Here is one related issue: https://github.com/Azure/autorest.csharp/issues/3440

ArcturusZhang commented:
Add back the null check for collections

ArcturusZhang commented:
Note - need to revert the changes on property bags since it is working right now

ArcturusZhang commented:
Hi @m-nash this is ready for review.

ArcturusZhang commented on test/CadlRanchProjects/type/property/nullable/src/Generated/Models/CollectionsByteProperty.Serialization.cs:
```diff
@@ -32,6 +32,11 @@ internal static CollectionsByteProperty DeserializeCollectionsByteProperty(JsonE
                 }
                 if (property.NameEquals("nullableProperty"u8))
                 {
+                    if (property.Value.ValueKind == JsonValueKind.Null)
```
I believe we should not do this.
We will get into this if when we get this json in response:
```
{ "nullableProperty": null }
```
If we want a model to serialize  into the json above, we will need to ensure the value of this collection is `null`. And we should do the same here when it is in the response, this makes the serialization and deserialization round trip.

Here are the model definition:
```csharp
public partial class Foo
{
    public Foo(IEnumerable<string> requiredNullableList)
    {
		RequiredNullableList = requiredNullableList?.ToList();
	}

	public IList<string> RequiredNullableList { get; set; }
}
```
If we have this code:
```csharp
var model = new Foo(null);
```
it will serialize into:
```json
{ "requiredNullableList": null }
```
And if you deserialize from the above json, you will get an instance of Foo, and its `RequiredNullableList` property is `null` again, something like:
```csharp
var model = new Foo(null);
var json = Serialize(model); // json is `{ "requiredNullableList": null }`
var deserializedModel = Deserialize(json);
Console.WriteLine(model.RequiredNullableList == deserializedModel.RequiredNullableList); // true
```
Same thing applies to optional list. If we have:
```csharp
var model = new Foo();
model.OptionalNullableList = null;
```
it serializes into
```json
{ "optionalNullableList": null }
``` 
and when we deserialize above json, we get an instance of `Foo` whose `OptionalNullableList` property is null.
If we have:
```csharp
var model = new Foo();
```
it serializes into nothing:
```json
{ }
```
and when we deserialize above json, we get an instance of `Foo` whose `OptionalNullableList` property is the ChangeTrackingList with `IsUndefined = true`.
I will add some more test cases in `Models-TypeSpec` to demo these.

ArcturusZhang replied:
We resolved this in offline discussion

ArcturusZhang commented on src/AutoRest.CSharp/Common/Input/CodeModelConverter.cs:
```diff
@@ -244,7 +244,8 @@ private InputModelType GetOrCreateModel(ObjectSchema schema, SchemaUsageProvider
                 DerivedModels: derived,
                 DiscriminatorValue: schema.DiscriminatorValue,
-                DiscriminatorPropertyName: schema.Discriminator?.Property.CSharpName());
+                DiscriminatorPropertyName: schema.Discriminator?.Property.CSharpName(),
```
This is very likely to be wrong - we should use serialized name of the property because that will never change.

ArcturusZhang replied:
Fixed in another PR

ArcturusZhang commented:
> @ArcturusZhang it appears that the changes are "no-op" meaning the functional effect is the same that we end up with an default changetrackinglist?

@m-nash 
This is how things work:
For optional lists, we will do this:
```csharp
Optional<IList<string>> optionalList = default;
foreach (var property in root.EnumerateObjects())
{
    if (property.NameEquals("optionalList"u8))
    {
        // this is the part handling null values
        if (property.Value.ValueKind == JsonValueKind.Null)
        {
			// we do nothing here so that the optional struct now is still "default"
			continue;
		}
    }
	// others
	return new Model(Optioanl.ToList(optionalList)); // when optionalList is default, this will return a changeTrackingList instance, and when it is not default, it returns that list instance (null included)
}
```
Code of Optional.ToList: https://github.com/Azure/autorest.csharp/blob/671929c1d9fbd74dc5bf53862bd1599ddd4af164/src/assets/Generator.Shared/Optional.cs#L73
When `optionalList` is default, `optionalList.HasValue` is false, therefore we return new ChangeTrackingList.

For required lists, we will do this:
```csharp
IList<string> requiredList = default; // which is `null`
foreach (var property in ...)
{
	if (property.NameEquals("requiredList"u8))
	{
        // this is the part handling null values
        if (property.Value.ValueKind == JsonValueKind.Null)
        {
			// we assign an instance of change tracking list here so that it is not null
			requiredList = new ChangeTrackingList<string>();
			continue;
		}
	}
	return new Model(requiredList); // no magic here
}
```

Hope this answers your question

ArcturusZhang commented:
Hi @m-nash I added a configuration `deserialize-null-collection-as-null-value` in this PR and it defaults to false.
When set to true, we will have previous behavior.
I have updated the GA libraries in the [corresponding regen PR](https://github.com/Azure/azure-sdk-for-net/pull/37553) by adding this flag to them, also I fixed issues in openai library, please also take a look.

The changes in generated files of openai I believe is expected, here is the corresponding typespec file: https://github.com/Azure/azure-rest-api-specs/blob/a89c3854f2d1cca8c996e00c97af3179ecb9dd72/specification/cognitiveservices/OpenAI.Inference/models/chat.completions.tsp#L196

ArcturusZhang commented:
@m-nash 
Last time decision on sync up meeting:
since changing the behavior of nullable collections is a behavioral breaking change, we should add a configuration to control that behavior, and I added a new configuration `deserialize-null-collection-as-null-value` which defaults to false (true corresponds to the old behavior).

Therefore in the regen PR, 
- There are quite a few RPs only have changes on autorest.md or tspConfig.yml without code changes, this is adding the config to stay on the old behavior to avoid behavior breaking changes.
- For those RPs with code changes, they are all in beta or preview therefore I did not add the new flag and breaking changes for them should be fine.
Other RPs might have API changes (for instance OpenAI) which is expected, because the initial purpose of this PR is fixing some issue around the required nullable properties. I added the corresponding typespec link in a comment in the regen to describe this change is expected.

lirenhe commented on src/AutoRest.CSharp/Common/Output/Models/Types/ModelTypeProviderFields.cs:
```diff
@@ -116,21 +116,41 @@ public ModelTypeProviderFields(InputModelType inputModel, TypeFactory typeFactor
+            if (property.IsDiscriminator)
+            {
+                // Discriminator properties should be writeable
+                return false;
```
Could you explain why we need the discriminator property to be writeable?

ArcturusZhang replied:
Sure

pshao25 commented on test/TestProjects/FirstTest-TypeSpec/FirstTest-TypeSpec.tsp:
```diff
@@ -111,6 +111,24 @@ model Thing {
+    requiredNullableList: int32[] | null;
+}
+
+@doc("A model with a few required nullable properties")
```
Why these not in cadl-ranch?

ArcturusZhang replied:
I do not know, maybe we could add some

ArcturusZhang replied:
We could add some new test cases in cadl ranch

AlexanderSher commented:
Can we please add a test that will do alias spread with required collections?

ArcturusZhang commented:
> Can we please add a test that will do alias spread with required collections?

Added

m-nash reviewed (approved):
@ArcturusZhang please create a follow up issue to reduce the granularity of the new flag `deserialize-null-collection-as-null-value`.  I imagine there is a small set of models / properties that have this issue and we don't want to proliferate it any further.

The granularity should probably be a list of `ModelName.PropertyName` to grandfather in.

ArcturusZhang commented:
> @ArcturusZhang please create a follow up issue to reduce the granularity of the new flag `deserialize-null-collection-as-null-value`. I imagine there is a small set of models / properties that have this issue and we don't want to proliferate it any further.
> 
> The granularity should probably be a list of `ModelName.PropertyName` to grandfather in.

issue created for this: https://github.com/Azure/autorest.csharp/issues/3613
